Lake/Build/Module.lean

```diff
@@ -136,79 +121,46 @@ def Module.recBuildLeanCore (mod : Module)
-`.c` file if the module is not lean-only or the requested artifact is `c`.
-Returns a build job producing the requested artifact.
--/
-def Module.recBuildLean (mod : Module) (art : LeanArtifact)
-: IndexBuildM (BuildJob (if art = .none then Unit else FilePath)) := do
-  let leanOnly := mod.isLeanOnly ∧ art ≠ .c
-  let modJob ← mod.recBuildLeanCore leanOnly
-  -- Save All Resulting Jobs & Return Requested One
-  store mod.leanBin.key modJob
-  let oleanJob ← modJob.bindSync fun _ depTrace =>
-    return (mod.oleanFile, mixTrace (← computeTrace mod.oleanFile) depTrace)
-  store mod.olean.key <| oleanJob
-  let ileanJob ← modJob.bindSync fun _ depTrace =>
-    return (mod.ileanFile, mixTrace (← computeTrace mod.ileanFile) depTrace)
-  store mod.ilean.key <| ileanJob
-  if h : leanOnly then
-    have : art ≠ .c := h.2
-    return match art with
-    | .none => modJob
-    | .olean => oleanJob
-    | .ilean => ileanJob
-  else
-    let cJob ← modJob.bindSync fun _  _ =>
-      return (mod.cFile, mixTrace (← computeTrace mod.cFile) (← getLeanTrace))
-    store mod.c.key cJob
-    return match art with
-    | .none => modJob
-    | .olean => oleanJob
-    | .ilean => ileanJob
-    | .c => cJob
+  extraDepJob.bindAsync fun _ _ => do
+  importJob.bindAsync fun _ importTrace => do
+  modDynlibsJob.bindAsync fun modDynlibs modTrace => do
+  externDynlibsJob.bindSync fun externDynlibs externTrace => do
+    let depTrace := importTrace.mix <| modTrace.mix externTrace
+    /-
+    Requirements:
+    * Lean wants the external library symbols before module symbols.
+    * Unix requires the file extension of the dynlib.
+    * For some reason, building from the Lean server requires full paths.
+      Everything else loads fine with just the augmented library path.
+    * Linux still needs the augmented path to resolve nested dependencies in dynlibs.
+    -/
+    let dynlibPath := libDirs ++ externDynlibs.filterMap (·.dir?) |>.toList
+    let dynlibs := externDynlibs.map (·.path) ++ modDynlibs.map (·.path)
+    let trace ← mod.buildUnlessUpToDate dynlibPath dynlibs depTrace
+    return ((), trace)
 
 /-- The `ModuleFacetConfig` for the builtin `leanBinFacet`. -/
 def Module.leanBinFacetConfig : ModuleFacetConfig leanBinFacet :=
-  mkFacetJobConfig (·.recBuildLean .none)
+  mkFacetJobConfig (·.recBuildLeanCore)
 
 /-- The `ModuleFacetConfig` for the builtin `oleanFacet`. -/
 def Module.oleanFacetConfig : ModuleFacetConfig oleanFacet :=
-  mkFacetJobConfig (·.recBuildLean .olean)
+  mkFacetJobConfigSmall fun mod => do
+    (← mod.leanBin.fetch).bindSync fun _ depTrace =>
+      return (mod.oleanFile, mixTrace (← computeTrace mod.oleanFile) depTrace)
 
 /-- The `ModuleFacetConfig` for the builtin `ileanFacet`. -/
 def Module.ileanFacetConfig : ModuleFacetConfig ileanFacet :=
-  mkFacetJobConfig (·.recBuildLean .ilean)
+  mkFacetJobConfigSmall fun mod => do
+    (← mod.leanBin.fetch).bindSync fun _ depTrace =>
+      return (mod.ileanFile, depTrace)
```

Comment:
The trace should be `mixTrace (← computeTrace mod.ileanFile) depTrace` like it originally was. Should it not?